entity-framework/core/miscellaneous/testing/index.md

```diff
@@ -1,16 +1,106 @@
+This illustrates the main trade off involved throughout these approaches: when is it appropriate to deviate from the production database system so as to improve test efficiency?
+
+Luckily, in this case the answer is quite easy: use local or on-premises SQL Server for developer testing.
+SQL Azure and SQL Server are extremely similar, so testing against SQL Server is usually a reasonable trade off.
+That being said, it is still wise to run tests against SQL Azure itself before going into production.
+ 
+### LocalDb 
+
+All the major database systems have some form of "Developer Edition" for local testing.
+SQL Server also also has a feature called [LocalDb](/sql/database-engine/configure-windows/sql-server-express-localdb?view=sql-server-ver15).
+The primary advantage of LocalDb is that it spins up the database instance on demand.
+This avoids having a database service running on your machine even when you're not running tests.
+
+LocalDb is not without it's issues:
+* It doesn't support everything that [SQL Server Developer Edition](/sql/sql-server/editions-and-components-of-sql-server-2016?view=sql-server-ver15) does.
+* It isn't available on Linux.
+* It can cause lag on first test run as the service is spun up.
+
+Personally, I've never found it a problem having a database service running on my dev machine and I would generally recommend using Developer Edition instead.
+However, it may be appropriate for some people, especially on less powerful dev machines.  
+
+## Approach 2: SQLite
+
+EF Core tests the SQL Server provider primarily by running it against a local SQL Server instance.
+These tests run tens of thousands of queries in a couple of minutes on a fast machine.
+This illustrates that using the real database system can be a performant solution.
+It is a myth that using some lighter-weight database is the only way to run tests quickly.
+
+That being said, what if for whatever reason you can't run tests against something close to your production database system?
+The next best choice is to use something with similar functionality.
+This usually means another relational database, for which [SQLite](https://sqlite.org/index.html) is the obvious choice.
+
+SQLite is a good choice because:
+* It runs in-process with your application and so has low overhead.
+* It uses simple, automatically created files for databases, and so doesn't require database management.
+* It has an in-memory mode that avoids even the file creation.
+
+However, remember that:
+* SQLite inevitability doesn't support everything that your production database system does.
+* SQLite will behave differently than your production database system for some queries.
+
+So if you do use SQLite for some testing, make sure to also test against your real database system.
+
+See [Testing with SQLite](xref:core/miscellaneous/testing/sqlite) for EF Core specific guidance. 
+
```

Comment:
Most developers use in-memory instead of relational for perf reasons, so we should gives some tips how to speed it up. Assuming most of the time is spent creating the database/schema:

- Write tests that don't change the database or are wrapped in a transaction that is rolled back after the test is finished.
- Clean the database data instead of recreating the schema. See [SqlServerDatabaseCleaner.cs](https://github.com/dotnet/efcore/blob/72639663d1e618c91385e1e57ce06530fa330450/test/EFCore.SqlServer.FunctionalTests/TestUtilities/SqlServerDatabaseCleaner.cs) for a generic way of how to do this.
- Create multiple copies of the database and assign one per test class to be able to run the tests in parallel

Reply:
Agreed. I have been sample tests and had similar thoughts. I think I'll create a new document for this to go alongside the updated docs for SQLite and in-memory.

Reply:
Will do this in another PR.